--- a/swift/stdlib/cmake/modules/AddSwiftStdlib.cmake	2025-07-11 17:20:43
+++ b/swift/stdlib/cmake/modules/AddSwiftStdlib.cmake	2025-07-11 17:23:01
@@ -2128,6 +2128,8 @@
         list(APPEND swiftlib_link_flags_all "-shared")
         # TODO: Instead of `lib${name}.so` find variable or target property which already have this value.
         list(APPEND swiftlib_link_flags_all "-Wl,-soname,lib${name}.so")
+        # Ensure compatibility with Android 15+ devices using 16KB memory pages.
+        list(APPEND swiftlib_link_flags_all "-Wl,-z,max-page-size=16384")
       endif()
 
       if (SWIFTLIB_BACK_DEPLOYMENT_LIBRARY)
